File: Eclipse Projects/webapps/TeleDirectJavaServer/src/com/teleDirect/Security.java
package com.teleDirect;

import java.sql.ResultSet;
import java.sql.SQLException;

import com.teleDirect.database.DatabaseProvider;
import com.teleDirect.server.Response;

/**
 * This class is responsible for looking after the security of the application.
 * It performs the following functions: login, logout & session validation.
 * @author AG041955
 *
 */
public class Security {

	/**
	 * Validates the login credentials from the database and 
	 * returns whether validation was successful or not.
	 * 
	 * @param username The username in String format
	 * @param password The password in String format
	 * @return A Response object containing whether it was 'success' or 'failure'.
	 * On successful login, the data in Repsonse contains the session details. 
	 */
	public Response login(String username, String password){
		String status = "", message = ""; 
		Object data = null;
		User curUser = null; 
		Session curSession = null;
		
		String sql = "SELECT name, priveleges FROM users WHERE username=\"" + username + "\" AND password=\"" + password + "\"";
		
		try {
			DatabaseProvider db = new DatabaseProvider();
			
			ResultSet result = db.executeQuery(sql);
			
			if (result == null)
			{
				throw new Exception("Sql query failed.");
			}
			
			int rowCount = 0;
			while(result.next()) {   // Move the cursor to the next row
	            String name = result.getString("name");
	            String priveleges = result.getString("priveleges");
	            
	            curUser = new User(username, name, priveleges);
	            ++rowCount;
			}
			
			db.closeDatabaseConnection(result);
			
			if(rowCount == 1){
				status = "success";
				message = "Welcome, " + curUser.getUsername();
				curSession = Session.start(curUser);
				data = (Object) curSession;
			}else if (rowCount == 0){
				status = "failure";
				message = "Wrong username/password";
			}else{
				throw new Exception("More than one username/password pair exist in the database.");
			}
			
		}catch (SQLException e) {
			status = "error";
			message = e.getMessage() + " We are unable to process your request. Please try again later.";
			e.printStackTrace();
		}catch (Exception e) {
			status = "error";
			message = e.getMessage() + " We are unable to process your request. Please try again later.";
			e.printStackTrace();
		}

		return new Response(status, message, data);
	}
	
	/**
	 * Logs out the user with the given username by ending the ongoing session.
	 * 
	 * @param username The username in String format
	 * @return A Response object containing whether it was 'success' or 'failure'.
	 */
	public Response logout(String username){
		String status = "", message = ""; 
		User user = new User(username, "", "");
		
		Session session = new Session("", user);
		boolean success = session.end();
		
		if(success){
			status = "success";
			message = username + " has logged out successfully.";
		}else{
			status = "failure";
			message = username + " could not be logged out.";
		}
		
		return new Response(status, message, null);
	}
	
	/**
	 * Validates the session details from the database and 
	 * returns whether validation was successful or not.
	 * 
	 * @param username The username in String format
	 * @param token The session token in String format
	 * @return A Response object containing whether it was 'success' or 'failure' 
	 * and an appropriate message.
	 */
	public Response validateSession(String username, String token){
		String status = "", message = ""; 
		
		String sql = "SELECT session_token AS token FROM users WHERE username=\"" + username + "\"" ;
		
		try {
			DatabaseProvider db = new DatabaseProvider();
			
			ResultSet rs = db.executeQuery(sql);
			String dbToken = "";
			int rowCount = 0;
			while(rs.next()) {   // Move the cursor to the next row
	            dbToken = rs.getString("token");
	            ++rowCount;
			}
			
			db.closeDatabaseConnection(rs);
			
			if(rowCount == 1){
				if(token.equals(dbToken)){
					status = "success";
					message = "Sesssion is valid.";
				}else{
					status = "failure";
					message = "Session not valid.";
				}
			}else if (rowCount == 0){
				status = "failure";
				message = "Username: \""+ username + "\" not found in database.";
			}else{
				throw new Exception("More than one username exist in the database.");
			}
			
		}catch (SQLException e) {
			status = "error";
			message = e.getMessage() + " We are unable to process your request. Please try again later.";
			e.printStackTrace();
		}catch (Exception e) {
			status = "error";
			message = e.getMessage() + " We are unable to process your request. Please try again later.";
			e.printStackTrace();
		}
		
		return new Response(status, message, null);
	}
}

File: Eclipse Projects/webapps/TeleDirectPhp/WebContent/scripts/php/Security.php
<?php

$file = basename(__FILE__ ).': ';

require_once 'Session.php';
require_once 'User.php';
require_once 'DatabaseProvider.php';

class Security{
	
	function login($username, $password){
		global $file;
		$response = null;
	
		error_log($file."username = ".$username.", password = ".$password."\n", 3, PHP_SCRIPT_PATH."debug.log");
		
		$dbProvider = new DatabaseProvider();
		$database = $dbProvider->getDatabase();
		try
		{
	
			error_log($file."Preparing SQL statement\n", 3, PHP_SCRIPT_PATH."debug.log");
	
			/*** prepare the select ***/
			$stmt = @$database->prepare('SELECT name, priveleges FROM users WHERE username = :username AND password = :password');
	
			/*** execute the prepared statement ***/
			@$stmt->execute(array(':username' => $username, ':password' => $password));
	
			error_log($file."Got results\n", 3, "debug.log");
			$numResults = $stmt->rowCount();
	
			error_log($file.$numResults." match found\n", 3, PHP_SCRIPT_PATH."debug.log");
	
			//echo $numResults ." results found\n";
			if($numResults == 1){
				$results = $stmt->fetch(PDO::FETCH_ASSOC);
	
				$user = new User($username, $results['name'], $results['priveleges']);
				$session = new Session($user);
				$session->start();
				
				$status = 'success';
				$message = "Welcome, ".$user->getName();
	
				$response =	array(
						'session' => $session->toArray(),
						'user' => $user->toArray()
				);
					
			}elseif ($numResults == 0){
				/** no user found **/
				$status = 'failure';
				$message = "Wrong username/password";
			}else{
				/** there is some problem **/
				throw new Exception("More than one username/password pair exist in the database.");
			}
		}
		catch(Exception $e)
		{
			$status = 'error';
			/*** if we are here, something has gone wrong with the database ***/
			$message = $e->getMessage() .' We are unable to process your request. Please try again later.';
		}
	
		return  json_encode(array(
				'status' => $status,
				'message' => $message,
				'data' => $response
		));
	}
	
	function logout($username){
		global $file;
		$response = null;
		try
		{
		
			$user = new User($username,'','');
			$session = new Session($user);
			$success = $session->end();
		
			if($success){
				
				$status = 'success';
				$message = $username. " has logged out successfully.";

			}else{
				$status = 'failure';
				$message = $username. " could not be logged out.";
			}
		}
		catch(Exception $e)
		{
			$status = 'error';

			/*** if we are here, something has gone wrong with the database ***/
			$message = $e->getMessage() .' We are unable to process your request. Please try again later.';
		}
		
		return  json_encode(array(
				'status' => $status,
				'message' => $message,
				'data' => $response
		));		
	}
	
	function valid($username, $token){
		global $file;
		
		$dbProvider = new DatabaseProvider();
		$database = $dbProvider->getDatabase();
		
		try
		{
		
			/*** prepare the select ***/
			$stmt = @$database->prepare('SELECT session_token FROM users WHERE username = :username');
			
			/*** execute the prepared statement ***/
			$success = @$stmt->execute(array(':username' => $username));
		
			if($success){
				
				$results = $stmt->fetch(PDO::FETCH_ASSOC);
				$dbToken = $results['session_token'];
				
				error_log($file."token = ".$token.", dbToken = ".$dbToken."\n", 3, PHP_SCRIPT_PATH."debug.log");
				if($token == $dbToken){
					$status = 'success';
					$message = "Session is valid.";
				}else{
					$status = 'failure';
					$message = "Session not valid.";
				}
			}
		}
		catch(Exception $e)
		{
			$status = 'error';
			
			/*** if we are here, something has gone wrong with the database ***/
			$message = $e->getMessage() .' We are unable to process your request. Please try again later.';
		}
		
		return  json_encode(array(
				'status' => $status,
				'message' => $message,
		));
	}
	
}


?>